### PR TITLE
Cleanups in loading of jar files

### PR DESCRIPTION
**Pull request by NathanJPhillips**
This includes adding a wrapper for the extract to file functionality to complement the extract to memory wrapper.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- N/A My commit message includes data points confirming performance improvements (if claimed).
- N/A My PR is restricted to a single feature or bugfix.
- N/A White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: jbmc/src/java_bytecode/mz_zip_archive.cpp [hunk 1]
```diff
@@ -73,13 +73,18 @@ size_t mz_zip_archivet::get_num_files()
...
-  const auto id=static_cast<mz_uint>(index);
-  std::vector<char> buffer;
-  buffer.resize(mz_zip_reader_get_filename(m_state.get(), id, nullptr, 0));
-  mz_zip_reader_get_filename(m_state.get(), id, buffer.data(), buffer.size());
-  // Buffer may contain junk returned after \0
-  const auto null_char_it=std::find(buffer.cbegin(), buffer.cend(), '\0');
-  return { buffer.cbegin(), null_char_it };
+  const auto id = static_cast<mz_uint>(index);
+  mz_uint name_size = mz_zip_reader_get_filename(m_state.get(), id, nullptr, 0);
+  if(name_size == 0)
```
Comment (smowton) [inline jbmc/src/java_bytecode/mz_zip_archive.cpp:78]: Braces round multi-line `if`

File: jbmc/src/java_bytecode/mz_zip_archive.cpp [hunk 2]
```diff
@@ -104,3 +104,19 @@ std::string mz_zip_archivet::extract(const size_t index)
...
   throw std::runtime_error("Could not extract the file");
 }
+
+void mz_zip_archivet::extract_to_file(
+  const size_t index,
+  const std::string &path)
+{
+  const auto id = static_cast<mz_uint>(index);
+  mz_zip_archive_file_stat file_stat;
+  if(mz_zip_reader_file_stat(m_state.get(), id, &file_stat) != MZ_TRUE)
```
Comment (smowton) [inline jbmc/src/java_bytecode/mz_zip_archive.cpp]: `file_stat` is unused

**Review by smowton — COMMENTED** [review]
I suspect you'll be asked to remove extract-to-file because it's unused

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 31cfb2a036eef1f803f260a99e9777dca4816239).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/112922881

**Comment by NathanJPhillips** [comment]
@smowton - happy to remove extract-to-file if you'd like - although there are other bits of functionality in JBMC that are only used by derivative projects. We can keep it for ourselves but it seemed like it might be useful to expose it for other potential users.

**Review by smowton — APPROVED** [review]
I'm happy with this as-is, just noting a common complaint re: committing untested / unused code :)

File: jbmc/src/java_bytecode/mz_zip_archive.cpp [hunk 3]
```diff
@@ -103,3 +103,16 @@ std::string mz_zip_archivet::extract(const size_t index)
   }
   throw std::runtime_error("Could not extract the file");
 }
+
+void mz_zip_archivet::extract_to_file(
```
Comment (peterschrammel) [inline jbmc/src/java_bytecode/mz_zip_archive.cpp:107]: That doesn't seem to match the PR title.

File: jbmc/src/java_bytecode/mz_zip_archive.cpp [hunk 4]
```diff
@@ -73,13 +73,17 @@ size_t mz_zip_archivet::get_num_files()
...
-  std::vector<char> buffer;
-  buffer.resize(mz_zip_reader_get_filename(m_state.get(), id, nullptr, 0));
-  mz_zip_reader_get_filename(m_state.get(), id, buffer.data(), buffer.size());
-  // Buffer may contain junk returned after \0
-  const auto null_char_it=std::find(buffer.cbegin(), buffer.cend(), '\0');
-  return { buffer.cbegin(), null_char_it };
+  const auto id = static_cast<mz_uint>(index);
+  mz_uint name_size = mz_zip_reader_get_filename(m_state.get(), id, nullptr, 0);
+  if(name_size == 0)
+    return {};  // Failure
```
Comment (tautschnig) [inline jbmc/src/java_bytecode/mz_zip_archive.cpp]: @NathanJPhillips clang-format would like to see one space removed.

**Comment by tautschnig** [comment]
@NathanJPhillips Please rebase now that #4729 is in to make CI happy.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 6b5dfb1d15e2591087658f33b7c01119ae0256ea).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/113796715

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 72936d15fd7aabd02bb537bf08a69bd785a5824f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/113895476

**Pull request merged**